File: android_dependencies/src/main/java/com/app/dependencies/data/dao/RealmLinkBookmarkDAO.kt
package com.app.dependencies.data.dao

import com.app.dependencies.data.dto.RealmLinkBookmarkDTO
import com.app.dependencies.data.dto.RealmTopicDTO
import com.app.shared.data.dao.LinkBookmarkDAO
import com.app.shared.data.dto.BookmarkDTO
import com.app.shared.data.dto.TopicDTO

class RealmLinkBookmarkDAO(override val instance: RealmProvider): RealmDAO, LinkBookmarkDAO {

    override fun getAll(): List<BookmarkDTO.LinkBookmarkDTO> {
        val realm = instance()
        val result = realm.where(RealmLinkBookmarkDTO::class.java).findAll()
        return realm.copyFromRealm(result).toList()
    }

    override fun insert(dto: BookmarkDTO.LinkBookmarkDTO) {
        val realm = instance()

        val topicDTO = dto.topic?.let {
            RealmTopicDTO(id = it.id, name = it.name)
        }

        val roomDTO = RealmLinkBookmarkDTO(
            id = dto.id,
            date = dto.date,
            title = dto.title,
            caption = dto.caption,
            icon = dto.icon,
            url = dto.url,
            topic = topicDTO
        )

        realm.beginTransaction()
        realm.insertOrUpdate(roomDTO)
        realm.commitTransaction()
    }

    override fun delete(bookmarkId: Int) {
        val realm = instance()
        val result = realm.where(RealmLinkBookmarkDTO::class.java).equalTo("id", bookmarkId).findAll()
        realm.beginTransaction()
        result.deleteAllFromRealm()
        realm.commitTransaction()
    }

    override fun replaceTopic(topicId: Int, withTopicDTO: TopicDTO) {
        val realm = instance()

        val topicResult = realm.where(RealmTopicDTO::class.java).equalTo("id", withTopicDTO.id).findFirst() ?: return
        val bookmarkResult = realm.where(RealmLinkBookmarkDTO::class.java).equalTo("topic.id", topicId).findAll() ?: return

        realm.beginTransaction()

        bookmarkResult.forEach {
            it.topic = topicResult
        }

        realm.commitTransaction()
    }
}

File: ios_app/ios_dependencies/ios_dependencies/Data/DAO/RealmLinkBookmarkDAO.swift
//
//  RealmLinkBookmarkDAO.swift
//  ios_dependencies
//
//  Created by Gabriel Coman on 14/04/2020.
//  Copyright © 2020 Gabriel Coman. All rights reserved.
//

import Foundation
import RememberShared
import Realm
import RealmSwift

public class RealmLinkBookmarkDAO: RealmDAO, LinkBookmarkDAO {
    
    let realm: Realm?
    
    public init(realm: Realm?){
        self.realm = realm
    }
    
    public func insert(dto__ dto: BookmarkDTOLinkBookmarkDTO) {
        try? realm?.write {
            realm?.add(dto.toObject(), update: Realm.UpdatePolicy.all)
        }
    }
    
    public func getAll() -> [BookmarkDTOLinkBookmarkDTO] {
        let result = realm?.objects(RealmLinkBookmarkObject.self)
        return result?.map { $0 }.map { $0.toDTO() } ?? []
    }
    
    public func delete(bookmarkId: Int32) {
        let result = realm?.objects(RealmLinkBookmarkObject.self).filter { $0.id == bookmarkId }
        
        guard let dto = result?.first else {
            return
        }
        
        try? realm?.write {
            realm?.delete(dto)
        }
    }
    
    public func replaceTopic(topicId: Int32, withTopicDTO: TopicDTO) {
        let topicResult = realm?.objects(RealmTopicObject.self).filter { $0.id == withTopicDTO.id }
        let bookmarkResult = realm?.objects(RealmLinkBookmarkObject.self).filter { $0.topic?.id == topicId }
        
        guard let defaultTopic = topicResult?.first, let bookmarks = bookmarkResult else {
            return
        }
        
        try? realm?.write {
            bookmarks.forEach {
                $0.topic = defaultTopic
            }
        }
    }
}
